File: examples/hello-extension/hooks.py
"""Lifecycle hooks for Hello Extension.

This module demonstrates how to implement Python lifecycle hooks
for VS Write extensions. Hooks allow extensions to react to
application events like project opening, section saving, etc.

Hook functions receive an ExtensionContext and a dict of arguments.
"""

from typing import Any, Dict
import logging

# Set up logging for debugging
logger = logging.getLogger(__name__)


def on_activate(ctx: Any, arguments: Dict[str, Any]) -> str:
    """Called when the extension is activated.

    Args:
        ctx: ExtensionContext with runtime access
        arguments: Empty dict for this hook

    Returns:
        Status message
    """
    logger.info("[HelloExtension] Activated!")
    return "Hello Extension activated successfully"


def on_deactivate(ctx: Any, arguments: Dict[str, Any]) -> str:
    """Called when the extension is deactivated.

    Args:
        ctx: ExtensionContext with runtime access
        arguments: Empty dict for this hook

    Returns:
        Status message
    """
    logger.info("[HelloExtension] Deactivated")
    return "Hello Extension deactivated"


def on_project_open(ctx: Any, arguments: Dict[str, Any]) -> str:
    """Called when a project is opened.

    Args:
        ctx: ExtensionContext with runtime access
        arguments: Dict containing 'project' with project data

    Returns:
        Status message
    """
    project = arguments.get("project", {})
    project_name = project.get("name", "Unknown")
    logger.info(f"[HelloExtension] Project opened: {project_name}")
    return f"Noted project open: {project_name}"


def on_project_close(ctx: Any, arguments: Dict[str, Any]) -> str:
    """Called when a project is closed.

    Args:
        ctx: ExtensionContext with runtime access
        arguments: Empty dict for this hook

    Returns:
        Status message
    """
    logger.info("[HelloExtension] Project closed")
    return "Project close acknowledged"


def on_section_save(ctx: Any, arguments: Dict[str, Any]) -> str:
    """Called after a section is saved.

    Args:
        ctx: ExtensionContext with runtime access
        arguments: Dict containing 'section' with section data

    Returns:
        Status message
    """
    section = arguments.get("section", {})
    section_title = section.get("title", "Unknown")
    content_length = len(section.get("content", ""))
    logger.info(f"[HelloExtension] Section saved: {section_title} ({content_length} chars)")
    return f"Tracked save: {section_title}"


def on_section_delete(ctx: Any, arguments: Dict[str, Any]) -> str:
    """Called after a section is deleted.

    Args:
        ctx: ExtensionContext with runtime access
        arguments: Dict containing 'sectionId'

    Returns:
        Status message
    """
    section_id = arguments.get("sectionId", "Unknown")
    logger.info(f"[HelloExtension] Section deleted: {section_id}")
    return f"Noted deletion: {section_id}"


def on_entity_change(ctx: Any, arguments: Dict[str, Any]) -> str:
    """Called when an entity is created, updated, or deleted.

    Args:
        ctx: ExtensionContext with runtime access
        arguments: Dict containing 'entity' with entity data

    Returns:
        Status message
    """
    entity = arguments.get("entity", {})
    entity_name = entity.get("name", "Unknown")
    entity_type = entity.get("type", "unknown")
    logger.info(f"[HelloExtension] Entity changed: {entity_name} ({entity_type})")
    return f"Tracked entity change: {entity_name}"

File: examples/starter-extension-lua/hooks.lua
-- =============================================================================
-- STARTER EXTENSION - LIFECYCLE HOOKS
-- =============================================================================
-- This file contains lifecycle hook implementations.
-- Hooks are called by VS Write at specific points in the application lifecycle.
--
-- Available hooks (enable in manifest.json):
--   on_activate       - Extension loaded
--   on_deactivate     - Extension unloaded
--   on_project_open   - Project opened (receives project info)
--   on_project_close  - Project closed
--   on_section_save   - Section saved (receives section)
--   on_entity_change  - Entity created/updated/deleted (receives entity)
--
-- Hook functions receive an 'args' table and must return:
--   { success = true, result = "..." }   - Success
--   { success = false, error = "..." }   - Failure
-- =============================================================================

--- Called when the extension is loaded
-- @param args table (empty)
-- @return table { success, result/error }
function on_activate(args)
    -- Perform initialization here
    -- Example: load configuration, set up state, etc.

    return {
        success = true,
        result = "Extension activated successfully"
    }
end

--- Called when the extension is unloaded
-- @param args table (empty)
-- @return table { success, result/error }
function on_deactivate(args)
    -- Perform cleanup here
    -- Example: save state, release resources, etc.

    return {
        success = true,
        result = "Extension deactivated"
    }
end

--- Called when a project is opened
-- @param args table { project_path, project_name }
-- @return table { success, result/error }
function on_project_open(args)
    local project_path = args.project_path or "unknown"
    local project_name = args.project_name or "unknown"

    -- Example: Initialize project-specific state
    -- You could load project preferences, scan files, etc.

    return {
        success = true,
        result = "Project opened: " .. project_name
    }
end

-- =============================================================================
-- ADDITIONAL HOOKS (uncomment and enable in manifest.json to use)
-- =============================================================================

--[[
--- Called when a project is closed
-- @param args table (empty)
-- @return table { success, result/error }
function on_project_close(args)
    return {
        success = true,
        result = "Project closed"
    }
end

--- Called when a section is saved
-- @param args table { section_id, section_title, content }
-- @return table { success, result/error }
function on_section_save(args)
    local section_id = args.section_id
    local section_title = args.section_title
    -- local content = args.content  -- The full section content

    -- Example: Auto-tag entities, validate content, etc.

    return {
        success = true,
        result = "Section saved: " .. section_title
    }
end

--- Called when an entity is created, updated, or deleted
-- @param args table { entity_id, entity_name, entity_type, action }
-- @return table { success, result/error }
function on_entity_change(args)
    local entity_id = args.entity_id
    local entity_name = args.entity_name
    local action = args.action  -- "create", "update", or "delete"

    -- Example: Update indexes, notify other extensions, etc.

    return {
        success = true,
        result = "Entity " .. action .. ": " .. entity_name
    }
end
]]--
